src/AIS.cpp

```diff
@@ -129,3 +136,147 @@ ais_base::VoyageInformation AIS::getVoyageInformation(ais::message_05 const& mes
+    base::Angle const& yaw,
+    base::Angle const& course_over_ground,
+    double speed_over_ground)
+{
+    if (!std::isnan(yaw.getRad())) {
+        return {
+            Eigen::Quaterniond(Eigen::AngleAxisd(yaw.getRad(), Eigen::Vector3d::UnitZ())),
+            ais_base::PositionCorrectionStatus::POSITION_CENTERED_USING_HEADING};
+    }
+    else if (!std::isnan(course_over_ground.getRad()) &&
+             speed_over_ground >= MIN_SPEED_FOR_VALID_COURSE) {
+        return {Eigen::Quaterniond(Eigen::AngleAxisd(course_over_ground.getRad(),
+                    Eigen::Vector3d::UnitZ())),
+            ais_base::PositionCorrectionStatus::POSITION_CENTERED_USING_COURSE};
+    }
+    else {
+        return {Eigen::Quaterniond::Identity(),
+            ais_base::PositionCorrectionStatus::POSITION_RAW};
+    }
+}
+
+/**
+ * @brief Transforms the sensor position relative to the vessel into the world frame
+ *
+ * @param sensor2vessel_pos The position of the sensor relative to the vessel
+ * @param vessel2world_ori The vessel's orientation in the world frame
+ *
+ * @return The sensor's position in the world frame
+ */
+base::Vector3d sensorToVesselInWorldPose(base::Vector3d const& sensor2vessel_pos,
+    Eigen::Quaterniond const& vessel2world_ori)
+{
+    base::Vector3d sensor2vessel_in_world_pos;
+    sensor2vessel_in_world_pos = vessel2world_ori * sensor2vessel_pos;
+
+    return sensor2vessel_in_world_pos;
+}
+
+/**
+ * @brief Converts a GPS position into UTM coordinates
+ *
+ * @param position The GPS position
+ * @param utm_converter The UTM converter
+ *
+ * @return A RigidBodyState in UTM coordinates
+ */
+base::samples::RigidBodyState convertGPSToUTM(ais_base::Position const& position,
+    gps_base::UTMConverter const& utm_converter)
+{
+    gps_base::Solution sensor2world_solution;
+    sensor2world_solution.latitude = position.latitude.getDeg();
+    sensor2world_solution.longitude = position.longitude.getDeg();
+
+    base::samples::RigidBodyState sensor2world_UTM;
+    sensor2world_UTM.position =
+        utm_converter.convertToUTM(sensor2world_solution).position;
+    sensor2world_UTM.time = base::Time::now();
+
+    return sensor2world_UTM;
+}
+
+/**
+ * @brief Converts a UTM position back to GPS coordinates in the world frame
+ *
+ * @param sensor2world_UTM The sensor position in UTM coordinates
+ * @param sensor2vessel_in_world_pos The sensor position in the world frame
+ * @param utm_converter The UTM converter
+ *
+ * @return A pair containing the latitude and longitude
+ */
+std::pair<base::Angle, base::Angle> convertUTMToGPSInWorldFrame(
+    base::samples::RigidBodyState const& sensor2world_UTM,
+    base::Vector3d const& sensor2vessel_in_world_pos,
+    gps_base::UTMConverter const& utm_converter)
+{
+    base::samples::RigidBodyState vessel2world_UTM;
+    vessel2world_UTM.position = sensor2world_UTM.position + sensor2vessel_in_world_pos;
```

Comment:
It is late and I am tired, so I can be very wrong but, shouldn't it be a minus? So it become `vessel2sensor_in_world_pos`
```suggestion
    vessel2world_UTM.position = sensor2world_UTM.position + (-1 * sensor2vessel_in_world_pos);
```

also I think the naming is a bit confusing, maybe place all math inside a single method would be less confusing

```cpp
// vessel and sensor orientation are the same 
auto vessel2sensor_pos = -sensor2vessel_pos;
vessel2world.position = sensor2world.position + sensor2world_ori * vessel2sensor_pos; 
```

Reply:
I think @wvmcastro is right. I looked it again

Reply:
> auto vessel2sensor_pos = -sensor2vessel_pos;

`sensor2vessel_pos` should actually be kept as `sensor2vessel_in_world_pos`, as @seixasxbr pointed out [here](https://github.com/tidewise/drivers-orogen-nmea0183/pull/10#discussion_r1925256554)

As for the naming, are you referring to the `_UTM` at the end? Because I'm working with conversions between UTM and GPS coordinates, I thought having the suffix `_UTM`/`_GPS` would clarify what type of coordinates I'm dealing with.

I also think you're right about the signal.

Reply:
sensor2world = (1,1)
sensor2vessel = (1,0)
vessel2world should be (0,1), not (2,1)
draw it in a 2D plane to help visualization

Reply:
> vessel2world_ori = identity sensor2world = (1,1) sensor2vessel_in_world = (1,0) vessel2world should be (0,1), not (2,1) draw it in a 2D plane to help visualization

What exactly is your point here? Please, be more objective in your comments.

Reply:
Nevermind